### PR TITLE
ARROW-7040: [C#] - System.Memory Span.CopyTo - Crashes

### PR DESCRIPTION
**Pull request by abbotware**
Work around for random crashes due to bug in Jitter / System.Memory
https://github.com/dotnet/coreclr/issues/27590

### CODE REVIEW

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs [hunk 1]
```diff
@@ -14,18 +14,65 @@
...
             MemoryMarshal.Cast<byte, T>(span);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static ReadOnlySpan<T> CastTo<T>(this ReadOnlySpan<byte> span)
             where T: struct =>
                 MemoryMarshal.Cast<byte, T>(span);
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this ReadOnlySpan<T> source, Span<T> target)
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: I don't think these methods should be `public`. We are going to want to remove them in the future when the underlying bug is fixed.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs [hunk 2]
```diff
@@ -14,18 +14,65 @@
...
+        public static void CopyToFix<T>(this Memory<T> source, Memory<T> target)
+        {
+            CopyToFix(source.Span, 0, target.Span, 0, source.Length);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(ReadOnlySpan<T> source, int sourceOffset, Span<T> target, int targetOffset, int length)
+        {
+            for (int i = 0; i < length; ++i)
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: This is a slower way of copying, and the underlying bug doesn't affect .NET Core.

We already multi-target for .NET Core. So can we only do the slow/workaround way when targeting `netstandard1.3`? And when we are on .NET Core, we can continue to use `Span.CopyTo`, which will perform faster than doing it one element at a time.

See https://github.com/apache/arrow/blob/e24825ac301d19d0b5f9f1dcdc404af98665a24f/csharp/src/Apache.Arrow/Apache.Arrow.csproj#L35-L40 for how we can switch out the implementation for different TFMs. You should be able to do the same thing for `SpanExtensions.cs`.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs [hunk 3]
```diff
@@ -14,18 +14,65 @@
...
         public static Span<T> CastTo<T>(this Span<byte> span)
             where T: struct =>
             MemoryMarshal.Cast<byte, T>(span);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static ReadOnlySpan<T> CastTo<T>(this ReadOnlySpan<byte> span)
             where T: struct =>
                 MemoryMarshal.Cast<byte, T>(span);
+
+
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: Can we add a comment here saying `TODO remove these methods once dotnet/coreclr#27590 is fixed`?

**Review by eerhardt — COMMENTED** [review]
Thanks for the fix @abbotware.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 2 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: @eerhardt  -  targeting `netcoreapp2.1` makes no sense (in my opinion)  it should just be various netstandardX.Ys and Net Framworks if it were to be multi-target

I did mention in the Jira I would make it multi-target, but the above code makes me think the multi-targeting is not setup correctly

Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: We multi-target to `netcoreapp2.1` so we can take advantage of newer APIs that don't exist in netstandard. That way when this library is used in an application running on .NET Core, it can use the newer APIs.

Your new code is a perfect example. Using the built-in `CopyTo` method is faster because it eventually calls `memcpy`, which will copy data faster than doing it one element at a time. Since there is no issue using the built-in `CopyTo` on .NET Core, it doesn't make sense to take the slower one-at-a-time route when running on .NET Core.

> but the above code makes me think the multi-targeting is not setup correctly

Can you explain what you mean here? Multi-targeting is set up correctly in this project. I followed the same pattern that we do in https://github.com/dotnet/runtime, which is to not use `#if` but instead split the different code into separate files, which makes it easier to reason about which code runs where. And it also aids in maintainability.

Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: What is an example of an API in NetCore that is not in NetStandard 2.0 or 2.1 ?

Maybe I am taking a purist approach - But i thought the generally accepted practice is to make libraries target NetStandard...   if you build an app you target NetCore or Framework and then implementation specific api's are use by the run-time.

Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: > What is an example of an API in NetCore that is not in NetStandard 2.0 or 2.1 ?

All the methods on Stream that accept Span / Memory parameters instead of arrays.

ex. https://github.com/apache/arrow/blob/249a12f131fa9e53aff035e5e5ad4ee6e38c3d05/csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs#L24-L27

These methods are not in `netstandard2.0` and below. They were added in `netstandard2.1`, but the only executable frameworks that supports `netstandard2.1` is .NET Core 3.0 and 3.1. So if we don't target `netcoreapp2.1`, then someone running on .NET Core 2.1 will get the slower implementation.

> But i thought the generally accepted practice is to make libraries target NetStandard

That is correct, and this library DOES target netstandard. However, there are places where we want to take advantage of faster APIs (like the Stream.Read method above accepting a Span). In those situations, we compile multiple times - once for `netstandard` and once for the lowest framework that has the faster API (in this case `netcoreapp2.1`).

Consumers don't see anything different. They can still run on .NET Framework or Xamarin just fine. But when they run on .NET Core 2.1+, the code built specifically for that framework is used.

Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: Right, i get that.. but I am curious how often that is even the case.   I suppose when i started looking at Arrow NetCore 3.0/3.1 and netstandard 2.0 / 2.1 were available as options so my views are somewhat anachronistic? 

FWIW - i don't see too much performance difference with this patch applied - And I am dealing with 15 gig file generation.

Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: >  I suppose when i started looking at Arrow NetCore 3.0/3.1 and netstandard 2.0 / 2.1 were available as options so my views are somewhat anachronistic?

We don't expect apps to move up to new versions of .NET immediately. .NET Core 3.0 has only been out for 3-4 months, and 3.1 (the long term support version) just came out last month. So not many applications are on it yet. .NET Core 2.1 will be supported at a minimum until August 21, 2021 (see https://github.com/dotnet/core/blob/master/microsoft-support.md). So keeping it running as fast as possible on .NET Core 2.1 is important.

> i don't see too much performance difference with this patch applied - And I am dealing with 15 gig file generation.

The code change might not be on the hot path for your scenario, but that doesn't mean it isn't for some other scenario. For example, if someone was using `ArrowBuffer.Builder.Append(ReadOnlySpan<T>)` and `Build` on a hot path.

> i'll setup the net472 target in an update to the PR

I don't understand this. I wasn't asking for anything to do with `net472`.

File: csharp/src/Apache.Arrow/Flatbuf/FlatBuffers/ByteBuffer.cs [hunk 4]
```diff
@@ -833,7 +834,7 @@ public int Put<T>(int offset, T[] x)
                 AssertOffsetAndLength(offset, numBytes);
                 // if we are LE, just do a block copy
 #if ENABLE_SPAN_T
-                MemoryMarshal.Cast<T, byte>(x).CopyTo(_buffer.Span.Slice(offset, numBytes));
+                MemoryMarshal.Cast<T, byte>(x).CopyToFix(_buffer.Span.Slice(offset, numBytes));
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Flatbuf/FlatBuffers/ByteBuffer.cs]: The Flatbuffer code is copied from https://github.com/google/flatbuffers/tree/master/net/FlatBuffers (which is the way https://google.github.io/flatbuffers/flatbuffers_guide_use_java_c-sharp.html says is how you consume flat buffer code).

I did a few checks, and the methods you are changing are not called in the Arrow library.

Can you leave the FlatBuffers code unchanged? That way when we update it from google/flatbuffers again, we don't have to worry about merge conflicts.

See https://github.com/apache/arrow/tree/master/csharp#updating-flatbuffers-code for more info.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 2 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: The patch I am supplying is to fix Span.CopyTo for a bug that only affects **.Net Framework** 

I thought creating a 3rd target for the arrow library (ie **net472**) is what you were proposing as a way to integrate this code so that only the target with the crash gets the fix - Or am I missing something?

Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: What I am proposing is to reuse the existing targets(`netstandard1.3` and `netcoreapp2.1`), and only doing the one-element-at-a-time copy on the `netstandard1.3` target, which is the one used when the consuming application is running on the .NET Framework. When an application is running on .NET Core (where there is no issue), it will use the Apache.Arrow assembly that is compiled for `netcoreapp2.1`. When compiling for `netcoreapp2.1`, we can use the built in `CopyTo` method.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs [hunk 5]
```diff
@@ -14,18 +14,65 @@
...
+            CopyToFix(source.Span, 0, target.Span, 0, source.Length);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(ReadOnlySpan<T> source, int sourceOffset, Span<T> target, int targetOffset, int length)
+        {
+            for (int i = 0; i < length; ++i)
+            {
+                target[targetOffset + i] = source[sourceOffset + i];
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: What happens when `target` and `source` overlap in memory? Does this handle that case?

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 2 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: ok - i thought you were arguing for minimal surface area of the affected target frameworks (since netstandard1.3 is much wider range of run times.. mono, Xamarin, etc) and the bug seems to only be in net framework. However... I am happy to put into netstandard1.3 as I am currently using a fork of arrow and this bug is the only reason I am not using the official package.

what do you think of file globing for the separate files?  (see here:)
https://github.com/abbotware/abbotware/blob/master/Common/Abbotware.Core/Abbotware.Core.csproj


the only issue doing this I have seen

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 5 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: @jkotas suggested pinning the spans before calling `CopyTo`, which should avoid the access violation on .NET Framework. This should avoid issues when the memory overlaps and be more efficient than one-element-at-a-time. Can you try doing that as an alternative fix?

```C#
CopyToFix()
{
    fixed (void* pFrom = MemoryMarshal.GetReference(spanFrom))
    fixed (void* pTo = MemoryMarshal.GetReference(spanTo))
    {
         spanFrom.CopyTo(spanTo);
    }
}
```

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 2 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: > what do you think of file globing for the separate files? 

As long as the Solution Explorer in VS works correctly with it, I would be fine with that solution.

Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: do you see both files now? I only see this with the current .csproj
image

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 1 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: made them internal

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 2 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: in either case - review the PR now as I did the split and added the globbing

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 5 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: I'll do some tests - I know the current implementation works.  With out the fix it would crash randomly with in the Arrow library - but almost immediately in the sample code i provided in: dotnet/coreclr#27590

File: .gitignore [hunk 6]
```diff
@@ -71,3 +71,10 @@ site/
 **/*.Rcheck/
 **/.Rhistory
 .Rproj.user
+/.vs
```
Comment (eerhardt) [inline .gitignore]: I don't think you need to change the `.gitignore` file with this change. Can you explain why this is being proposed here?

If there is a C# specific pattern that needs to be ignored (like `.vs`), there is a `csharp/.gitignore` file for the C# specific patterns - but it already has `.vs`.

File: csharp/src/Apache.Arrow/Apache.Arrow.csproj [hunk 7]
```diff
@@ -32,10 +32,11 @@
     </EmbeddedResource>
   </ItemGroup>
 
+
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Apache.Arrow.csproj]: (nit) inserting this blank line is unnecessary.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs [hunk 8]
```diff
@@ -14,16 +14,19 @@
 // limitations under the License.
 
 using System;
+using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace Apache.Arrow
 {
-    public static class SpanExtensions
+    public  static partial class SpanExtensions
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: (nit) double spaces between `public` and `static`

File: csharp/src/Apache.Arrow/Flatbuf/FlatBuffers/ByteBuffer.cs [hunk 9]
```diff
@@ -41,6 +41,7 @@
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 using System.Text;
+using Apache.Arrow;
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Flatbuf/FlatBuffers/ByteBuffer.cs]: Can you revert this change as well?

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.netstandard1.3.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            CopyToFix(source.Span, 0, target.Span, 0, source.Length);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void CopyToFix<T>(ReadOnlySpan<T> source, int sourceOffset, Span<T> target, int targetOffset, int length)
+        {
+            for (int i = 0; i < length; ++i)
+            {
+                target[targetOffset + i] = source[sourceOffset + i];
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Extensions/SpanExtensions.netstandard1.3.cs]: Are we guaranteed that `target` and `source` will never overlap in memory?

If they do, this is not going to work correctly. If the beginning of `target` is at the end of `source`, this code will mutate the end of `source` before it reads the last elements of `source`. Thus the wrong data will be copied to `target`.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.netstandard1.3.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void CopyToFix<T>(ReadOnlySpan<T> source, int sourceOffset, T[] target, int targetOffset, int length)
```
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Extensions/SpanExtensions.netstandard1.3.cs]: Why do we need this method? Won't the above method that takes `Span<T> target` be enough, since `T[]` is implicitly convertible to `Span<T>`?

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 5 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: > I know the current implementation works

It may work for the cases you tested. But it doesn't work when `target` and `source` overlap in memory. If the beginning of `target` and the end of `source` are the same memory location, by the time the loop gets to the end of `source`, you have overwritten the original values in `source`. Thus the wrong information gets copied to `target`.

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.netstandard1.3.cs (hunk 11 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.netstandard1.3.cs]: it might no longer be needed (I removed the usage in flat buffers) - I created different overloads to get the compilation to work when I did a global replace of Span.CopyTo - If i created it, it was being used at some point for compliation

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 5 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: if they overlap I would expect that to be a program bug (although I can see where that would be useful)  if Span.CopyTo handles overlaps, then yes my simple approach is not correct as a temp copy is needed

File: .gitignore (hunk 6 above)
Comment (abbotware) [reply .gitignore]: if you click "Open in Visual Studio" on github - the root folder is opened and folders are created automatically... Any one else that does that will probably see those folders get created as well - so its just annoying to see those all the time..

File: csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs (hunk 5 above)
Comment (abbotware) [reply csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs]: (this is very clear)
https://docs.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.copyto?view=netcore-3.1

(the remarks should probably match the the above wording)
https://docs.microsoft.com/en-us/dotnet/api/system.span-1.copyto?view=netcore-3.1

I'll test out @jkotas suggestion and change the patch... as that seems like it solves a few problems at once

**Comment by wesm** [comment]
What is the status of this? Needs a rebase

**Comment by wesm** [comment]
Following up again on this since 10 weeks have passed

**Comment by wesm** [comment]
I'm closing this since it has gone stale but maybe someone can pick it up in the future

**Comment by eerhardt** [comment]
Just a follow up here - a fix has been checked into .NET Framework 4.7 and 4.8 that makes this change no longer necessary. According to https://github.com/dotnet/runtime/issues/13701#issuecomment-670722879 the fix is scheduled to be deployed in September.
